### PR TITLE
WIP: Use common beam instead of beam averaging

### PR DESCRIPTION
**Pull request by e-koch**
Remove beam averaging and force operations to use the common beam. This depends on improvements to the common beam algorithm in radio-beam (https://github.com/radio-astro-tools/radio-beam/pull/81).

* [x] Remove current usage of beam averaging
* [ ] Generalize identification of bad beams to difference in areas instead of deviation from median. *requires changes in radio-beam*
* [ ] Toggle to enforce common beam operations vs. allow small deviations. With dask, would add spatial convolution operations before computing and/or introduce a "hidden" intermediate dump of a zarr file with the convolved cube (may be unfeasible, we'll see)
* [ ] Cache common beam somewhere if `goodbeam_mask` doesn't change

### CODE REVIEW

**Comment by e-koch** [comment]
@astrofrog - Have you seen this `InvocationError` before? https://travis-ci.org/github/radio-astro-tools/spectral-cube/jobs/721172051 I can't reproduce it locally.

**Comment by astrofrog** [comment]
I’ll try and take a look shortly!

**Comment by e-koch** [comment]
@astrofrog Fixed it (somehow)! I think a file wasn't being closed when the test failed.

File: spectral_cube/base_class.py [hunk 1]
```diff
@@ -554,30 +561,53 @@ def identify_bad_beams(self, threshold, reference_beam=None,
...
+                      "This is because beam averaging is not the correct operation. Please use"
+                      " `cube.set_common_beam` in its place to compute the common beam. The "
+                      "common beam can then be accessed with `cube.common_beam`.",
+                      DeprecationWarning)
+
+        self.set_common_beam(threshold, mask=mask, warn=warn)
+
+    def set_common_beam(self,
+                        threshold=None, mask='compute', warn=False,
+                        combeam_kwargs={}):
```
Comment (astrofrog) [inline spectral_cube/base_class.py:578]: I'm not a fan of the ``combeam`` abbreviation - maybe just ``common_beam_kwargs``, or why not simply have ``**kwargs`` and pass them through to the common beam algorithm? Is there a risk of overlap? Since the main purpose of the method is to actually call the common beam algorithm, just using ``**kwargs`` would seem more natural to me?

Comment (e-koch) [reply spectral_cube/base_class.py]: Agreed. I've switched to `**kwargs` here.

**Comment by e-koch** [comment]
Tests on travis are giving an `InvocationError` again when one test using a dask cube fails and I think the file doesn't get closed.

Here's the failure I get locally:
```
___________________________________________________________________ test_unmasked_channels[True-data_vda] ___________________________________________________________________

filename = PosixPath('/tmp/pytest-of-eric/pytest-60/test_unmasked_channels_True_da0/vda.fits'), use_dask = True

    @pytest.mark.parametrize('filename', ['data_vda', 'data_vda_beams'],
                             indirect=['filename'])
    def test_unmasked_channels(filename, use_dask):
    
        cube, data = cube_and_raw(filename, use_dask=use_dask)
    
        # Add a mask to the cube
        mask = np.ones(cube.shape, dtype=bool)
        mask[0] = False
        cube = cube.with_mask(mask)
    
>       np.testing.assert_equal(cube.unmasked_channels, mask.any(axis=(1, 2)))

spectral_cube/tests/test_spectral_cube.py:2570: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
spectral_cube/utils.py:18: in wrapper
    self._cache[(func, args)] = func(self, *args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = DaskSpectralCube with shape=(4, 3, 2) and unit=K and chunk size (4, 3, 2):
 n_x:      2  type_x: RA---SIN  unit_x: deg...094 deg:   29.935209 deg
 n_s:      4  type_s: VOPT      unit_s: km / s  range:     -321.215 km / s:    -317.350 km / s

    @property
    @cached
    def unmasked_channels(self):
        if isinstance(self._data, da.Array):
>           return self._compute(da.any(self._mask_include, axis=(1, 2)))
E           AttributeError: 'DaskSpectralCube' object has no attribute '_mask_include'

spectral_cube/base_class.py:353: AttributeError
========================================================================== short test summary info ==========================================================================
FAILED spectral_cube/tests/test_spectral_cube.py::test_unmasked_channels[True-data_vda] - AttributeError: 'DaskSpectralCube' object has no attribute '_mask_include'

```

The non-varying resolution cube isn't getting the mask set at an earlier stage. But the test with a VRSC is fine.

**Comment by e-koch** [comment]
@astrofrog @keflavich -- ready for review

File: docs/beam_handling.rst [hunk 2]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+Finding and convolving to a common resolution
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When applying spectral operations on a
+`~spectral_cube.spectral_cube.VaryingResolutionSpectralCube`, it is often
+necessary to first convolve the data to have a common beam. To retain the
+maximum amount of spatial information, the data can be convolved to the
+smallest common beam that all beams in
+`~spectral_cube.spectral_cube.VaryingResolutionSpectralCube.beam` can be
+deconvolved from.
```
Comment (keflavich) [inline docs/beam_handling.rst:81]: this is... pretty hard to parse.  Try to rephrase?  Maybe just end at "smallest common beam"?  Isn't it "smallest beam that all beams can be convolved to" anyway?

File: docs/beam_handling.rst [hunk 3]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+
+Finding the smallest common beam and convolution operations are described
```
Comment (keflavich) [inline docs/beam_handling.rst:83]:
```suggestion
Tools for finding the smallest common beam and convolution operations are described
```

File: docs/beam_handling.rst [hunk 4]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+in detail in :doc:`smoothing`.
+
+
+Small beam variations: Do I need to convolve first?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Formally, operations on multiple spectral channels require the data to have a common
+resolution. In practice, data cubes with a high-spectral resolution will have the beam
```
Comment (keflavich) [inline docs/beam_handling.rst:91]:
```suggestion
resolution. In practice, data cubes with a high spectral resolution will have the beam
```

File: docs/beam_handling.rst [hunk 5]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+the beams will be ignored, a warning will be printed describing this approximation, and
+the spectral operation will be computed.
+
+The beam threshold can be changed from the default 0.01 with::
+
+    >>> vrsc_cube.beam_threshold = 0.02   # doctest: +SKIP
+
+.. note::
+
+    Setting `vrsc_cube.beam_threshold = 1.0` will allow all beam variations without raising an error. Large values for the beam threshold should not be used for scientific results.
```
Comment (keflavich) [inline docs/beam_handling.rst:111]: I think this is not true. You can have more than 100% variation in beam size, then this will still fail.  I've had this happen when there was a 7m beam in 1 channel and real 12m beams in the others

File: docs/beam_handling.rst [hunk 6]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+
+For most spectral-line data cubes covering a single spectral line, the fine resolution
+and small frequency range will often allow the above approximation to work.
+However, if you are working with wideband data, this approximation will not hold and you
+will need to convolve to a common beam, or extract spectral slabs from the data cube.
```
Comment (keflavich) [inline docs/beam_handling.rst:116]:
```suggestion
will need to convolve to a common beam or extract spectral slabs from the data cube.
```

**Review by keflavich — CHANGES_REQUESTED** [review]
some minor changes.

I just want to make sure: the default behavior is not to convolve to a common beam, is it?  I want to be sure we're not triggering massive convolution operations without warning the user, since that tends to crash computers.

**Comment by e-koch** [comment]
No, no auto trigger of a whole cube convolution. An error is raised if the beams deviate too much from the common beam, similar to what we already have with the median/avg beam.

File: docs/beam_handling.rst (hunk 2 above)
Comment (e-koch) [reply docs/beam_handling.rst]: Fixed.

File: docs/beam_handling.rst (hunk 5 above)
Comment (e-koch) [reply docs/beam_handling.rst]: Fixed. Noted that 1.0 and larger allow for large beam variations (not "all").

File: docs/smoothing.rst [hunk 7]
```diff
@@ -29,12 +29,31 @@ will be different for each slice.
...
 
-    common_beam = cube.beams.common_beam()
-    new_cube = cube.convolve_to(common_beam)
+Many spectral operations on a cube with a varying resolution will require first
+convolving to a common resolution. You may want to retain the finest spatial
+resolution possible by finding the smallest common beam that encloses all beams
+in the cube. To do this, we can use
+:meth:`~spectral_cube.VaryingResolutionSpectralCube.set_common_beam`::
+
+    cube.set_common_beam()
```
Comment (astrofrog) [inline docs/smoothing.rst:39]: Is there a reason why we need to make a stateful change on the cube and set ``common_beam``? We could instead do something like:

```
common_beam = cube.find_common_beam()
cube.convolve_to(common_beam)
```

? Having it stateful means it could get out of sync if e.g. some channels are masked and so on. In general we might want to avoid stateful changes to cubes if we can?

With ``find_common_beam`` above one could even find a common beam with different settings and compare them without having to make a stateful change each time.

File: spectral_cube/base_class.py [hunk 8]
```diff
@@ -507,8 +514,10 @@ def identify_bad_beams(self, threshold, reference_beam=None,
 
         Parameters
         ----------
-        threshold : float
-            Fractional threshold
+        threshold : float, optional
+            The fractional difference between beam major, minor, and pa to
```
Comment (astrofrog) [inline spectral_cube/base_class.py:518]: maybe spell out position angle?

File: spectral_cube/base_class.py [hunk 9]
```diff
@@ -688,38 +741,79 @@ def _check_beam_areas(self, threshold, mean_beam, mask=None):
...
-    def mask_out_bad_beams(self, threshold, reference_beam=None,
+            if raise_error or self.strict_beam_match:
+                raise ValueError(f"{strictmessage}{errormessage}\nConvolve to a common beam before applying any"
+                                 " spectral operation.")
+
+            else:
+                warnings.warn(errormessage)
+                warnings.warn(strictmessage)
+                warnings.warn("Convolution to a common beam will be triggered at an intermediate level."
+                              " To avoid this step, first convolve the spectral-cube to a common beam size.")
```
Comment (astrofrog) [inline spectral_cube/base_class.py:785]: spectral-cube -> spectral cube or SpectralCube (with the hyphen is the package name?)

**Review by astrofrog — CHANGES_REQUESTED** [review]
Looks good overall! A question below about API, in particular I think it would be cleaner to return a common beam when it is computed than store it in a stateful way on the cube.

File: docs/smoothing.rst (hunk 7 above)
Comment (e-koch) [reply docs/smoothing.rst:39]: The only issue is that the common beam operation can be slow for cubes with many (>1e3) channels. And baked into the checks for deviations is the need to check for variation from the common beam, where we had the average beam before. That's our only check (currently) for whether to ignore small beam changes for operations along the spectral axis.

Allowing a stateful change could also work if we force the beam masking to not be stateful (which I think it can be now)

File: docs/beam_handling.rst [hunk 10]
```diff
@@ -53,19 +53,122 @@ Multi-beam cubes
...
+of 0.01; see above).
+However, the comparison here is to the **median** major and minor axis rather than
+the smallest common beam used above.
+This is because bad beams are identified as outliers in the set of beams.
+
+To mask the channels with bad beams, use
+`~spectral_cube.spectral_cube.VaryingResolutionSpectralCube.mask_out_bad_beams`.
+
+
+    >>> masked_vrsc_cube = vrsc_cube.mask_out_bad_beams()  # doctest: +SKIP
```
[the remark this answers is not in the document]
Comment (e-koch) [reply docs/beam_handling.rst:159]: `mask_out_bad_beams` was in before this PR. I could a deprecation warning here? I agree with `with_bad_beams_masked`

File: spectral_cube/base_class.py (hunk 8 above)
Comment (e-koch) [reply spectral_cube/base_class.py]: :+1:

File: docs/beam_handling.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (astrofrog) [reply docs/beam_handling.rst:159]: Yes I think a deprecation warning would make sense, @keflavich do you agree?

[the remark this answers is not in the document]
Comment (keflavich) [reply docs/beam_handling.rst:159]: yep, :+1: to deprecating.

[the remark this answers is not in the document]
Comment (e-koch) [reply docs/beam_handling.rst]: Done!

**Comment by e-koch** [comment]
* [ ] Tests for loading varying-res cubes with common beam computed (`compute_commonbeam=True`)
* [ ] Docs for loading with `compute_commonbeam=True`
* [ ] Test for `with_bad_beams_masked` (replacing state change in `identify_bad_beams`)
* [ ] Update docs for `with_bad_beams_masked`
* [ ] Test for new `convolve_to_commonbeam`
* [ ] Add docs to convolution operations for `convolve_to_commonbeam`

**Comment by keflavich** [comment]
Discussion today: @e-koch will try splitting up this PR to avoid state-change related issues.

Maybe need to remove median(beams) operation
